pyro/distributions/mixture.py

```diff
@@ -102,14 +102,17 @@ def expand(self, batch_shape):
 
     def sample(self, sample_shape=torch.Size()):
         mask = self.mask.expand(sample_shape + self.batch_shape) if sample_shape else self.mask
-        result = self.component0.sample(sample_shape)
-        result[mask] = self.component1.sample(sample_shape)[mask]
+        mask = mask.reshape(mask.shape + (1,) * len(self.event_shape))
+        result = torch.where(mask,
+                             self.component1.sample(sample_shape),
+                             self.component0.sample(sample_shape))
         return result
 
     def rsample(self, sample_shape=torch.Size()):
         mask = self.mask.expand(sample_shape + self.batch_shape) if sample_shape else self.mask
-        result = self.component0.rsample(sample_shape)
-        result[mask] = self.component1.rsample(sample_shape)[mask]
```

Comment:
@ordabayevy Could you also reshape the mask here?

Reply:
@fehiepsi ok, thanks. Reshaping in the log_prob seems to be unnecessary.

Reply:
I think it is required. The `mask` shape there also does not include `event_shape`.

Reply:
including it there fails the test, presumably because the output of the log_prob has the shape of `sample_shape + batch_shape` without the `event_shape`

Reply:
Opps, you are right!!!